Replace runtime.Rich* implicits on primitives by direct extension methods.

sjrd opened the pull request:
It has always been weird that the public API of primitive types, enriched by `Predef` depends on the `Rich*` classes, which are in the not-really-public `runtime` package.

Moreover, the design of those classes was overabstracted, with consequences on performance (unnecessary boxing) and quality of the API (including methods that do not make sense on some types). To mitigate that, the individual `Rich*` classes redefined some (but not all) of the methods, defeating the abstraction.

We solve both issues with a simple solution: define all those methods as simple `extension` methods. We do this directly in the companion objects of the primitive types.

---

This would be a required step before we do anything about #23824.

sjrd committed:
Replace runtime.Rich* implicits on primitives by direct extension methods.

It has always been weird that the public API of primitive types,
enriched by `Predef` depends on the `Rich*` classes, which are
in the not-really-public `runtime` package.

Moreover, the design of those classes was overabstracted, with
consequences on performance (unnecessary boxing) and quality of
the API (including methods that do not make sense on some types).
To mitigate that, the individual `Rich*` classes redefined some
(but not all) of the methods, defeating the abstraction.

We solve both issues with a simple solution: define all those
methods as simple `extension` methods. We do this directly in the
companion objects of the primitive types.

soronpo commented:
The problem is that extension methods are not equivalent to implicit classes. Without solving https://contributors.scala-lang.org/t/relaxed-extension-methods-sip-54-are-not-relaxed-enough/6585/1 this change will break everywhere that has methods in the same name in scope.

sjrd commented:
> The problem is that extension methods are not equivalent to implicit classes/definition extensions. Without solving https://contributors.scala-lang.org/t/relaxed-extension-methods-sip-54-are-not-relaxed-enough/6585/1 this change will break everywhere that has methods in the same name in scope.

I don't see how that problem appears when defining extensions for *monomorphic* types. You wouldn't redefine another extension of `toHexString` for `Int` (or if you do, it's on you), so you won't have a clash like that. With polymorphic types there are reasons to define separate extensions for `Foo[Int]` and `Foo[String]`, but for a monomorphic type that's not the case.

soronpo commented:
> > The problem is that extension methods are not equivalent to implicit classes/definition extensions. Without solving https://contributors.scala-lang.org/t/relaxed-extension-methods-sip-54-are-not-relaxed-enough/6585/1 this change will break everywhere that has methods in the same name in scope.
> 
> I don't see how that problem appears when defining extensions for _monomorphic_ types. You wouldn't redefine another extension of `toHexString` for `Int` (or if you do, it's on you), so you won't have a clash like that. With polymorphic types there are reasons to define separate extensions for `Foo[Int]` and `Foo[String]`, but for a monomorphic type that's not the case.

You are right. I though the implementation used a typeclass approach + extension methods. My mistake.

sjrd commented on library/src/scala/Byte.scala:
```diff
@@ -485,5 +487,89 @@ object Byte extends AnyValCompanion {
+      *
+      * @param end The final bound of the range to make.
+      */
+    def until(end: Byte): NumericRange.Exclusive[Byte] = NumericRange(self, end, 1)
```
This PR is strictly a refactoring. Adding `infix` anywhere should be done separately.

tewecske commented:
Will this change be available in 3.3.x LTS?

sjrd commented:
No.

sjrd commented:
I had a look at the failures mentioned above. One is an easy fix (an omission, really). Some are expected source or macro breakages (like in parboiled, whose macro cannot recognize `to` and `until` ranges anymore). Some are more annoying.

Given everything else that is happening in 3.8.0, I don't think it's wise to subject our users to one *more* thing. But if you disagree, I can rebase and fix the missing thing, and then it's one more thing in the 3.8.0 ballot. :man_shrugging:

hamzaremmal commented on library/src/scala/Boolean.scala:
```diff
@@ -138,5 +138,27 @@ object Boolean extends AnyValCompanion {
+      * - `0` if `this == that`
+      */
+    def compare(that: Boolean): Int = java.lang.Boolean.compare(self, that)
+
```
Also, all of these methods comes from the fact that `RichBoolean` inherits `scala.math.Ordered`.
I understand that we are porting what exists but at the same time, it's weird that we don't utilize the fact that `Boolean` is an `Ordering`.

sjrd replied:
You still get `compareTo` from auto-boxing `Boolean` to `java.lang.Boolean`.

hamzaremmal reviewed (dismissed):
Some observations here. I see that some extensions are missing like `compareTo`, `doubleValue`, `self`. Does this PR relies on the fact that we will fallback to `Rich*` via the implicit conversion? If yes, it should clearly state that. If we don't, does that mean that these methods will not be available in source anymore? That sets precedence to removing things in source but not in the binary.
On the other hand, I see that this PR added `to` and `until` on `Byte` which didn't use to exist in the `Rich*` world. Since this is a *refactoring* PR, we shouldn't have additions. I'm requesting changes as I want to have an answer to these questions before approving.

sjrd commented on library/src/scala/Byte.scala:
Same. These exist on `java.lang.Number`, and *a fortiori* on its subclasses, so you get them with auto-boxing.

hamzaremmal commented on library/src/scala/Short.scala:
```diff
@@ -484,5 +486,117 @@ object Short extends AnyValCompanion {
   implicit def short2double(x: Short): Double = x.toDouble
-}
 
+  extension (self: Short) {
```
> Do we want to reintroduce them as @deprecated?

I would say no but are we breaking any of our guarantees by doing so?

hamzaremmal commented:
It is indeed concerning but it is not as bad as we expected (broken test suite and infra). The Scala.js test, as written compiles with Scala 3.7.3 but it does not generalize. In this snippet, `x` compiles but `y` doesn't.
```scala
val x = (Byte.MinValue to Byte.MaxValue) // ok (what Scala.js test does)

object Foo:
  val Min: Byte = 0
  val Max: Byte = 10

val y = (Foo.Min to Foo.Max) // doesn't compile with Scala 3.7.3
```

AST after typer.

```scala
-- [E008] Not Found Error: test.scala:8:17 -------------------------------------
8 |val y = (Foo.Min to Foo.Max) // doesn't compile with Scala 3.7.3
  |         ^^^^^^^^^^
  |        value to is not a member of Byte.
  |        An extension method was tried, but could not be fully constructed:
  |
  |            intWrapper(Foo.Min)
[[syntax trees at end of                     typer]] // test.scala
package <empty> {
  final lazy module val Foo: Foo = new Foo()
  final module class Foo() extends Object() { this: Foo.type =>
    val Min: Byte = 0
    val Max: Byte = 10
  }
  final lazy module val test$package: test$package = new test$package()
  final module class test$package() extends Object() {
    this: test$package.type =>
    val x: scala.collection.immutable.Range.Inclusive = intWrapper(-128).to(127)
    val y:
      
        <error value to is not a member of Byte.
An extension method was tried, but could not be fully constructed:

    intWrapper(Foo.Min)>
        
     = Foo.Min.to(Foo.Max)
  }
}

1 error found
```

Gedochao commented:
I'll just link the issue here, so that the connection is easy to find.
The blocker seems to be:
- https://github.com/scala/scala3/issues/24321

Gedochao commented:
@sjrd I guess this is worth a rebase, since we're entering 3.10?

sjrd committed:
Remove the workarounds for `to/until` on `Byte/Short`s.

The code that they enable was only working with compile-time
constants in the first place. That is a deeper issue, which is
captured in #24321.

This change impacts the presentation compiler. One test progresses,
and one regresses.

sjrd commented:
@Gedochao Rebased.

It still breaks Parboiled in the community build, since it has a macro that cannot recognize the new tree shape. What do we do about that?

SethTisue commented:
> It still breaks Parboiled in the community build, since it has a macro that cannot recognize the new tree shape. What do we do about that?

@xuwei-k I see you've been doing the maintenance on parboiled, is it something you'd be interested in fixing?

xuwei-k commented:
- https://github.com/sirthias/parboiled2/pull/708

sjrd commented:
Yeah, this of is green now. Thanks a lot @xuwei-k for the parboiled changes!

@Gedochao this is ready for a real review now.

Note that I'll be on vacation for the next 3 weeks, so I might not be super responsive.

noti0na1 commented on library/src/scala/Float.scala:
```diff
@@ -401,5 +401,103 @@ object Float extends AnyValCompanion {
+    def sign: Float = java.lang.Math.signum(self)
+
+    /** Returns the signum of `this`. */
+    @deprecated("signum does not handle -0.0f or Double.NaN; use `sign` method instead", since = "2.13.0")
```
Should this be `Float.NaN`?

noti0na1 commented on library/src/scala/Int.scala:
The scaladoc still says:
```
 *  There is an implicit conversion from [[scala.Int]] => [[scala.runtime.RichInt]]
 *  which provides useful non-primitive operations.
```
, which should be updated with this PR.

noti0na1 commented on library/src/scala/Byte.scala:
```diff
@@ -14,6 +14,8 @@ package scala
 
 import scala.language.`2.13`
 
+import scala.collection.immutable.NumericRange
```
Unused import? also in `Short` and `Double`

noti0na1 reviewed (changes requested):
Otherwise, LGTM

sjrd commented:
Thanks for the review. Updated.
